// @sina/koa-config/lib/baseConfig.js
'use strict';

const config = {
    /**
     * 应用名称
     * @type {String}
     */
    PROJECT_NAME: '',

    /**
     * 应用域名
     * @type {String}
     */
    PROJECT_HOST: '',

    /**
     * 应用监听端口
     * @type {array}
     */
    PORT: [3000],

    /**
     * session secret
     * @type {String}
     */
    sessionSecret: '',

    /**
     * 允许POST的域
     * @type {Array}
     */
    acceptReferer: ['sina.com.cn', 'weibo.com', 'sina.cn', 'weibo.cn'],

    /**
     * 项目根目录
     * @type {String}
     */
    ROOT_PATH: '',

    /**
     * 项目配置目录
     * @type {String}
     */
    CONFIG_PATH: '',

    /**
     * APP目录
     * @type {String}
     */
    APP_PATH: '',

    /**
     * 项目模型目录
     * @type {String}
     */
    MODEL_PATH: '',

    /**
     * 项目静态文件目录
     * @type {String}
     */
    STATIC_PATH: '',

    /**
     * 核心文件目录，必填
     * @type {String}
     */
    LIB_PATH: '',

    /**
     * 模板目录
     * @type {String}
     */
    TEMPLATE_PATH: '',

    /**
     * 开启自动适配到wap模板
     * 1 开启
     * 0 关闭，default
     * @type {Number}
     */
    vt: 0,

    /**
     * 存放wap模板的目录，必须放在views下边，目录需要手动创建
     * @type {String}
     */
    VT_DIR: 'vt_4',

    /**
     * 模板配置
     * @type {Object}
     */
    render_config: {
        root: '', //模板路径
        autoescape: true, // 默认开启转移
        cache: 'memory', // Send false to disable
        ext: 'html' // 默认模板后缀
    },

    /**
     * MYSQL
     * @type {Object}
     */
    DATABASE: {
        debug: false,
        master: {
            host: '',
            port: 3306,
            database: '',
            user: '',
            password: '',
            charset: 'UTF8_GENERAL_CI',
            connectionLimit: 10, // 数据库操作为连接池操作，主库最大连接数根据实际情况设置
            connectTimeout: 150, // 连接超时，单位milliseconds
            dateStrings: true,
            supportBigNumbers: true // 有BIGINT and DECIMAL columns的时候，需要打开
        },
        slave: {
            host: '',
            port: 3306,
            database: '',
            user: '',
            password: '',
            charset: 'UTF8_GENERAL_CI',
            connectionLimit: 50, // 数据库操作为连接池操作，从库最大连接数根据实际情况设置
            connectTimeout: 150, // 连接超时，单位milliseconds
            dateStrings: true,
            supportBigNumbers: true // 有BIGINT and DECIMAL columns的时候，需要打开
        }
    },

    /**
     * MC
     * @type {Object}
     */
    MEMCACHED: {
        server: '', // [ '192.168.0.102:11211', '192.168.0.103:11211' ]
        options: {
            debug: false,
            namespace: '', // 默认前缀
            timeout: 100,
            reconnect: 2000,
            retries: 3,
            failures: 3,
            retry: 1000,
            remove: true
        }
    },

    /**
     * REDIS
     * @type {Object}
     */
    REDIS: {
        master: {
            host: '',
            port: 6379,
            options:{
                connect_timeout: 30000,
                max_attempts: 10,
                prefix: '',
                auth_pass: ''
            }
        },
        slave: {
            host: '',
            port: 6379,
            options:{
                connect_timeout: 3000,
                max_attempts: 10,
                prefix: '',
                auth_pass: ''
            }
        }
    },

    /**
     * s3
     * @type {Object}
     */
    S3: {},

    /**
     * error code
     * @type {Object}
     */
    LOG_CONFIG: {

        /**
         * DB
         * @type {Number}
         */
        ERROR_MODEL_ID_DB: 1,

        /**
         * HTTP
         * @type {Number}
         */
        ERROR_MODEL_ID_HTTP: 2,

        /**
         * SINA STORAGE SERVICE
         * @type {Number}
         */
        ERROR_MODEL_ID_S3: 3,

        /**
         * USER
         * @type {Number}
         */
        ERROR_MODEL_ID_USER: 4,

        /**
         * MEMCACAHED
         * @type {Number}
         */
        ERROR_MODEL_ID_MC: 5,

        /**
         * NODEJS
         * @type {Number}
         */
        ERROR_MODEL_ID_NODE: 6,

        /**
         * DEFAULT
         * @type {Number}
         */
        ERROR_MODEL_ID_DEFAULT: 7,

        /**
         * CRONTAB
         * @type {Number}
         */
        ERROR_MODEL_ID_QUEUE: 8,

        /**
         * PAY
         * @type {Number}
         */
        ERROR_MODEL_ID_PAY: 9,

        /**
         * REDIS
         * @type {Number}
         */
        ERROR_MODEL_ID_REDIS: 10,

        /**
         * 详细错误信息
         * @type {Array}
         */
        codeName: {
            c90100: '请求源不允许',
            c90101: '请求方法不允许，框架只允许get或post',
            c90102: 'Controller类中方法为基类方法不能使用',
            c90103: 'Controller类中不存在调用的方法',
            c90104: 'Warning错误',
            c90105: 'Fatal error错误',
            c90106: '脚本运行超默认上限',
            c90107: 'model类名冲突',
            c90200: '找不到对应route信息，请查看config/Routephp配置',
            c90201: '未指定state',
            c90202: '未指定action',
            c90203: 'delUrlParams函数缺少params参数',
            c90204: 'addUrlParams函数缺少params参数',
            c90205: '没有找到控制器类',
            c90300: '请求方法不允许，主库操作必须为post',
            c90301: '数据库返回非资源',
            c90302: 'insert或replace或insertOnDuplicate中inser_value或replace_value传参错误',
            c90303: 'insertOnDuplicate中update_keys参数在insert_value中不存在',
            c90304: 'insertOnDuplicate中update_keys参数无有效字段',
            c90305: 'update中update_value传参错误',
            c90306: 'update中where条件错误',
            c90307: 'delete中where条件错误',
            c90308: '字段在配置文件中未定义',
            c90309: '字段在配置文件中禁止修改',
            c90310: '字段值不符合在配置文件定义的类型',
            c90311: '数据库连接失败',
            c90312: 'sql不能为空',
            c90313: '传参不符合拼接规范，无法正确翻译sql语句',
            c90314: '数据库连接超过默认时间',
            c90315: '数据库操作错误',
            c90320: '数据库基类方法不存在',
            c90400: '服务器没有安装curl扩展',
            c90401: '页面抓取请求url缺失',
            c90402: 'url参数错误',
            c90403: '请求连续20次失败',
            c90404: 'CURL内部错误信息',
            c90405: 'http异常code',
            c90406: '重定向超预定次数',
            c90500: 'MC无法连接',
            c90501: 'MC连接超过默认时间',
            c90502: 'MC操作异常',
            c90600: 'Redis连接失败',
            c90601: 'Redis操作失败',
            c90602: 'Redis连接超过指定时间',
            c90603: 'Redis认证失败',
            c90604: 'Redis未配置host和port',
            c90700: '请指定有效上传文件',
            c90701: '请设置文件的mime type',
            c90702: '上传失败',
            c90703: '上传失败',
            c90704: '其他错误',
            c90705: 'S3配置不存在',
            c90706: 'S3认证配置缺失',
            c90707: 'sinaEdge删除缓存失败',
            c90708: 'sinaEdge预热缓存失败',
            c90709: '删除失败',
            c90800: '邮件发送失败',
            c90900: 'SAE不支持追加写入操作',
            c91000: '指定开关不存在',
            c91100: 'XML解析错误',
            c91200: '队列异常',
            c91201: '队列超时',
        },
    },

    /**
     * log4js config
     * @type {Object}
     */
    // LOG4JS_CONFIG: {
    //     development: {
    //         appenders: [
    //             {
    //                 type: 'console',
    //                 category: 'console',
    //             },
    //             {
    //                 type: 'dateFile',
    //                 absolute: true,
    //                 filename: '',
    //                 pattern: '_info-yyyy-MM-dd.log',
    //                 alwaysIncludePattern: true,       // 文件名是否始终包含占位符
    //                 category: 'info',
    //             },
    //             {
    //                 type: 'dateFile',
    //                 absolute: true,
    //                 filename: '',
    //                 pattern: '_error-yyyy-MM-dd.log',
    //                 alwaysIncludePattern: true,       // 文件名是否始终包含占位符
    //                 category: 'error',
    //             },
    //             {
    //                 type: 'console',
    //                 category: 'debugInfo',
    //             },
    //             {
    //                 type: 'console',
    //                 category: 'debugError',
    //             },
    //         ],
    //         replaceConsole: true,    //替换console.log
    //         levels: {
    //             console: 'ALL',
    //             info: 'ALL',
    //             error:  'ALL',
    //             debugInfo:  'TRACE',
    //             debugError:  'WARN',
    //         },
    //     },
    //     test: {
    //         appenders: [
    //             {
    //                 type: 'console',
    //                 category: 'console',
    //             },
    //             {
    //                 type: 'dateFile',
    //                 absolute: true,
    //                 filename: '',
    //                 pattern: '_info-yyyy-MM-dd.log',
    //                 alwaysIncludePattern: true,       // 文件名是否始终包含占位符
    //                 category: 'info',
    //             },
    //             {
    //                 type: 'dateFile',
    //                 absolute: true,
    //                 filename: '',
    //                 pattern: '_error-yyyy-MM-dd.log',
    //                 alwaysIncludePattern: true,       // 文件名是否始终包含占位符
    //                 category: 'error',
    //             },
    //             {
    //                 type: 'console',
    //                 category: 'debugInfo',
    //             },
    //             {
    //                 type: 'console',
    //                 category: 'debugError',
    //             },
    //         ],
    //         replaceConsole: true,    //替换console.log
    //         levels: {
    //             console: 'ALL',
    //             info: 'TRACE',
    //             error:  'WARN',
    //             debugInfo:  'TRACE',
    //             debugError:  'WARN',
    //         },
    //     },
    //     production: {
    //         appenders: [
    //             {
    //                 type: 'console',
    //                 category: 'console',
    //             },
    //             {
    //                 type: 'dateFile',
    //                 absolute: true,
    //                 filename: '',
    //                 pattern: '_info-yyyy-MM-dd.log',
    //                 alwaysIncludePattern: true,       // 文件名是否始终包含占位符
    //                 category: 'info',
    //             },
    //             {
    //                 type: 'dateFile',
    //                 absolute: true,
    //                 filename: '',
    //                 pattern: '_error-yyyy-MM-dd.log',
    //                 alwaysIncludePattern: true,       // 文件名是否始终包含占位符
    //                 category: 'error',
    //             },
    //             {
    //                 type: 'console',
    //                 category: 'debugInfo',
    //             },
    //             {
    //                 type: 'console',
    //                 category: 'debugError',
    //             },
    //         ],
    //         replaceConsole: true,    //替换console.log
    //         levels: {
    //             console: 'ALL',
    //             info: 'TRACE',
    //             error:  'WARN',
    //             debugInfo:  'TRACE',
    //             debugError:  'WARN',
    //         },
    //     },
    // },
};

module.exports = config;
